packages/jsii-diff/lib/classes-ifaces.ts

```diff
@@ -0,0 +1,203 @@
+import reflect = require('jsii-reflect');
+import log4js = require('log4js');
+import { Analysis, FailedAnalysis, isSuperType } from './type-analysis';
+import { ComparisonContext, shouldInspect } from './types';
+
+const LOG = log4js.getLogger('jsii-diff');
+
+/**
+ * Compare two class types
+ *
+ * We require that all stable properties and methods on the original are
+ * present on the new type, and that they match in turn.
+ */
+export function compareReferenceType<T extends reflect.ReferenceType>(original: T, updated: T, context: ComparisonContext) {
+  if (original.isClassType() && updated.isClassType()) {
+    if (updated.abstract && !original.abstract) {
+      context.mismatches.report(original, 'has gone from non-abstract to abstract');
+    }
+
+    // JSII assembler has already taken care of inheritance here
+    if (original.initializer && updated.initializer) {
+      compareMethod(original, original.initializer, updated.initializer, context);
+    }
+  }
+
+  if (original.docs.isSubclassable && !updated.docs.isSubclassable) {
+    context.mismatches.report(original, 'has gone from @subclassable to non-@subclassable');
+  }
+
+  for (const [origMethod, updatedMethod] of memberPairs(original, original.allMethods, updated, context)) {
+    compareMethod(original, origMethod, updatedMethod, context);
+  }
+
+  for (const [origProp, updatedProp] of memberPairs(original, original.allProperties, updated, context)) {
+    compareProperty(original, origProp, updatedProp, context);
+  }
+
+  // You cannot have added abstract members to the class/interface, as they are
+  // an added burden on potential implementors.
+  //
+  // Only for types that are explicitly marked as intended to be subclassed by customers.
+  if (subclassableType(original)) {
+    noNewAbstractMembers(original, updated, context);
+  }
+}
+
+export function compareStruct(original: reflect.InterfaceType, updated: reflect.InterfaceType, context: ComparisonContext) {
+  // We don't compare structs here; they will be evaluated for compatibility
```

Comment:
looks weird. wouldn't it make sense to just have an argument that says if this is an input/output struct? [and again I am going back to the intuition that this is an inherent property of a struct]